### PR TITLE
sql: primitive to compute merge ordering

### PR DESCRIPTION
**Pull request by RaduBerinde**
Adding a primitive that can be used to find the matching ordering for two sides
of a join.

### CODE REVIEW

**Comment by arjunravinarayan** [comment]
Expand the commit message to make clearer the goal - we need this to propagate orderings past a potential candidate merge, so that we can make better planning decisions when we have nested Joins (i.e. selecting ordering of which join to do first).

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 225 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotkbqkzpyzNEalkuCv:-KotkbqkzpyzNEalkuCw:bceae6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L225)):*
> ```Go
> // in colA/colB. Specifically column i in the resulting ordering refers
> // to the column that contains colA[i] from A and colB[i] from B.
> func computeMergeOrdering(a, b orderingInfo, colA, colB []int) sqlbase.ColumnOrdering {
> ```

why does this take `orderingInfo` but return a `ColumnOrdering`? Shouldn't it return an `orderingInfo` as well?

---

*[pkg/sql/ordering.go, line 235 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotmZuXFe5jit7iwXPM:-KotmZuXFe5jit7iwXPN:b-8hxgl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L235)):*
> ```Go
> 	var result sqlbase.ColumnOrdering
> 
> 	used := make([]bool, len(colA))
> ```

`used` needs a comment explaining what information its capturing (and why).

---

*[pkg/sql/ordering.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotnnJN6-DBSci03C_s:-KotnnJOv6Z9V68-F6Xu:b8bdrdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L259)):*
> ```Go
> 	}
> 
> 	for ordA, ordB := a.ordering, b.ordering; ; {
> ```

This really needs a comment block explaining the algorithm being used here, this is non-intuitive and hard to follow.

---

*[pkg/sql/ordering.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-Kotr25pbUKCdM6fw5Eg:-Kotr25pbUKCdM6fw5Eh:b-xycki1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L259)):*
> ```Go
> 	}
> 
> 	for ordA, ordB := a.ordering, b.ordering; ; {
> ```

My girlfriend also asks why the variables are named `aDone` and `bDone` but the other variables are of the form `ordA` and `ordB`. She would prefer `doneA` and `doneB` to maintain consistency.

---

*[pkg/sql/ordering.go, line 263 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-Kotolz7CKbDY045X0DG:-Kotolz7CKbDY045X0DH:bhgrig4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L263)):*
> ```Go
> 		// See if the first column in the orderings matches.
> 		if !aDone && !bDone {
> 			i, okA := eqMapA[ordA[0].ColIdx]
> ```

I don't understand what's happening here.

---

*[pkg/sql/ordering.go, line 301 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotpMDHkUYYEKgC6Kn4:-KotpMDHkUYYEKgC6Kn5:b-tf7lrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L301)):*
> ```Go
> 		break
> 	}
> 	// TODO(radu): if we consumed an ordering and it is ".unique" (i.e. at most
> ```

typo ".unique"

---

*[pkg/sql/ordering.go, line 303 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotpPwwY78jSNL__QU3:-KotpPwwY78jSNL__QU4:b-f6pxtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L303)):*
> ```Go
> 	// TODO(radu): if we consumed an ordering and it is ".unique" (i.e. at most
> 	// one row has the same set of values on those columns), then we can assume
> 	// the remainder of that ordering
> ```

incomplete sentence?

---

*[pkg/sql/ordering_test.go, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotqlbayD6sqdYtZH8D:-Kotqlbb-uICvez28MZu:bf2xmm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L275)):*
> ```Go
> 	asc := encoding.Ascending
> 	desc := encoding.Descending
> 	testCases := []struct {
> ```

I'm starting to think these style test cases could use a string field `propertyTested`, which briefly describes the test case (a subset of b, b subset of a, a = b, ...). This helps understand, and on test failure will be more helpful as well.

---

*[pkg/sql/ordering_test.go, line 328 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotqO4bfgI_5GVRgArT:-KotqO4bfgI_5GVRgArU:b3k7u4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L328)):*
> ```Go
> 		{
> 			a: orderingInfo{
> 				ordering: sqlbase.ColumnOrdering{o(1, asc)},
> ```

how does this test case work? There's no 2nd column in `a`.

---

*[pkg/sql/ordering_test.go, line 370 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-Kotqcwd3cn2799BCWIK:-Kotqcwd3cn2799BCWIL:b8e0t1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L370)):*
> ```Go
> 			colA:     []int{1, 2},
> 			colB:     []int{3, 4},
> 			expected: sqlbase.ColumnOrdering{o(0, asc)},
> ```

why isn't this one `o(0, asc), o(1, desc)`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-:-KotkDSq8FuYiiTh_vXR:bx6l5tz)*

**Comment by knz** [comment]
I agree with Arjun, the new function in `ordering.go` must be prepended by a comment which puts the algorithm into many more words with several examples.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-KovvLwSeX0BbJDAIBnc:bkyxapq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
> ```Go
> 			if _, ok := b.exactMatchCols[colB[i]]; ok {
> 				used[i] = true
> 				result = append(result, sqlbase.ColumnOrderInfo{ColIdx: i, Direction: encoding.Ascending})
> ```

The direction is not necessarily ascending. It depends how we do the merge.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-:-KovwuqprmpIU2Q1fGeW:bcqrof)*

**Comment by RaduBerinde** [comment]
Responded to some things, will clean it up according to your comments.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 225 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotkbqkzpyzNEalkuCv:-Kow9yxYs4uehAXLI4Wr:b-ys7s9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L225)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

why does this take `orderingInfo` but return a `ColumnOrdering`? Shouldn't it return an `orderingInfo` as well?
</blockquote></details>

The extra stuff in `orderingInfo` is not useful here (we can't have exact match columns or uniqueness). This function is meant to generate a ColumnOrdering that can be used to interleave the rows.
We will have a separate function for generating an `orderingInfo` for the join output (which is more complicated and requires extending the `orderingInfo` stuff because of equal and merged columns). I will clarify in the comment.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-KowAQv73Cd1vZopgmLB:b-rvwioz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The direction is not necessarily ascending. It depends how we do the merge.
</blockquote></details>

This function can return multiple valid results. The assumption is that the merge will be done according to this result, in which case it will hold on the merged output.

But you are right that this is an opportunity - we could merge either way and one ordering might be better than the other. We may want to use this during expand plan and pass a desired ordering too. I'll add a TODO.

---

*[pkg/sql/ordering_test.go, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotqlbayD6sqdYtZH8D:-Kow8JY7dAaiGGmbNuGr:bn8t9xs) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L275)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

I'm starting to think these style test cases could use a string field `propertyTested`, which briefly describes the test case (a subset of b, b subset of a, a = b, ...). This helps understand, and on test failure will be more helpful as well.
</blockquote></details>

In most cases it's not easy to come up with a short string that conveys enough information.. I think if you are debugging a case you will need to look at it and understand it anyway.

---

*[pkg/sql/ordering_test.go, line 328 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotqO4bfgI_5GVRgArT:-Kow8hr4yMQarIliDYeI:bcuj6cr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L328)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

how does this test case work? There's no 2nd column in `a`.
</blockquote></details>

unique=true means that the values for column 1 are distinct; so an ordering of 1+ is also equivalent to all orderings 1+, <whatever>+/-

---

*[pkg/sql/ordering_test.go, line 370 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-Kotqcwd3cn2799BCWIK:-Kow9ZObCQKEeHec0Z6h:bam3833) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L370)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

why isn't this one `o(0, asc), o(1, desc)`?
</blockquote></details>

because an ordering 3+,5-,4- does not imply an ordering of 3+,4-, example
```
3 5 4         3 4
-------------------
1 2 2         1 2
1 2 1    ->   1 1
1 1 3         1 3
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-:-KowB4-K9gVFIGt3u2aS:b-6901ww)*

**Comment by RaduBerinde** [comment]
Ok, updated the change with more comments and better commit message. I've kind of muddled two things together - determining if a merge-join can be used (and what ordering the merge-join must use internally), and determining the output ordering of a join. The latter is a more complicated problem that isn't tackled in this change; I clarified this.

---

Review status: 0 of 2 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/ordering.go, line 235 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotmZuXFe5jit7iwXPM:-KowgfwXVBTBtU7MBdqt:b-e75ul1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L235)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

`used` needs a comment explaining what information its capturing (and why).
</blockquote></details>

Removed, I think it was a left-over.

---

*[pkg/sql/ordering.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotnnJN6-DBSci03C_s:-KowhR9fXB7auHfbfA1t:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L259)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

This really needs a comment block explaining the algorithm being used here, this is non-intuitive and hard to follow.
</blockquote></details>

Added.

---

*[pkg/sql/ordering.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-Kotr25pbUKCdM6fw5Eg:-KowhSEo_BY8aB_cuNQz:bpny5nc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L259)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

My girlfriend also asks why the variables are named `aDone` and `bDone` but the other variables are of the form `ordA` and `ordB`. She would prefer `doneA` and `doneB` to maintain consistency.
</blockquote></details>

Tell her it's done.

---

*[pkg/sql/ordering.go, line 263 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-Kotolz7CKbDY045X0DG:-KowiWzGrvHS-BEfw9s2:bxy1ieb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L263)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

I don't understand what's happening here.
</blockquote></details>

Should  be clearer now.

---

*[pkg/sql/ordering.go, line 303 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotpPwwY78jSNL__QU3:-KowdEwNPAJjDro_T-Id:bomdoi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L303)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

incomplete sentence?
</blockquote></details>

Removed, this will apply when we determine the output ordering of a merge-join.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-:-KowidIItb7sMGKd6J4I:bevavbz)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 2 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/ordering_test.go, line 328 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotqO4bfgI_5GVRgArT:-Kowk7S3eDtRekY33QeG:b-hlhpd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L328)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

unique=true means that the values for column 1 are distinct; so an ordering of 1+ is also equivalent to all orderings 1+, <whatever>+/-
</blockquote></details>

that should read all orderings `1+, <whatever>+/-`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997)*

**Comment by arjunravinarayan** [comment]
:lgtm_strong: Thank you for the comments. It is now exceptionally well documented.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 239 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-Kp0grBFpXvMFnGKH8UQ:-Kp0grBFpXvMFnGKH8UR:bnyg2rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L239)):*
> ```Go
> 	// First, find any merged columns that are exact matches in both sources. This
> 	// means that in each source, this column only sees one value. Note that this
> 	// value can be different for the two sources, so the merged column is not
> ```

Why remove this last part of the comment? I found it useful.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-Kp0h0K3uirBLKqUI6_z:b-pbei4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This function can return multiple valid results. The assumption is that the merge will be done according to this result, in which case it will hold on the merged output.

But you are right that this is an opportunity - we could merge either way and one ordering might be better than the other. We may want to use this during expand plan and pass a desired ordering too. I'll add a TODO.
</blockquote></details>

Perhaps we need a `Direction: encoding.Either` type, which can be be resolved later by the planner.

---

*[pkg/sql/ordering_test.go, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotqlbayD6sqdYtZH8D:-Kp0hGbnbppvytSstm3W:b2nhktp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L275)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

In most cases it's not easy to come up with a short string that conveys enough information.. I think if you are debugging a case you will need to look at it and understand it anyway.
</blockquote></details>

OK.

---

*[pkg/sql/ordering_test.go, line 328 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotqO4bfgI_5GVRgArT:-Kp0i8GHl8jpum7o1uF7:b-9755fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L328)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

that should read all orderings `1+, <whatever>+/-`
</blockquote></details>

Thank you.

---

*[pkg/sql/ordering_test.go, line 370 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-Kotqcwd3cn2799BCWIK:-Kp0hSZZk-df8MlVXC84:b2qy1vw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L370)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

because an ordering 3+,5-,4- does not imply an ordering of 3+,4-, example
```
3 5 4         3 4
-------------------
1 2 2         1 2
1 2 1    ->   1 1
1 1 3         1 3
```
</blockquote></details>

Today I learned! Thank you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-:-Kp0iG5UM7RuXvykM-38:b-qrfrl0)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 239 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-Kp0grBFpXvMFnGKH8UQ:-Kp13ftagWOB40oOmi5t:bla2rw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L239)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Why remove this last part of the comment? I found it useful.
</blockquote></details>

It's not true anymore :)  The old function tried to be more general - it talked about general merging of rows. The new function talks about equality between columns. If they are equal, and both have a single value on all rows, that value has to be the same.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-Kp14NZWbBY6JAvJdEZ0:b-nkvuwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Perhaps we need a `Direction: encoding.Either` type, which can be be resolved later by the planner.
</blockquote></details>

An "either" direction is the same concept as the "exact match column" in `orderingInfo`. Perhaps we do want to return a more detailed answer in the form of an `orderingInfo`, but it's a bit hard to describe the semantics accurately. I will add a TODO for now.

---

*[pkg/sql/ordering_test.go, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotqlbayD6sqdYtZH8D:-Kp15t-Xa4jXfADDHj1T:b4dp2mo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L275)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

OK.
</blockquote></details>

I will take another look at this. The current state of things is also annoying in that when test `XX/15` fails you have to count through the test cases to get to the 15th. It would be much easier if there was a simpe name you could grep (even if it doesn't convey much information in many cases).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/ordering_test.go, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotqlbayD6sqdYtZH8D:-Kp1HTgKeOnvrBdxMhYT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L275)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I will take another look at this. The current state of things is also annoying in that when test `XX/15` fails you have to count through the test cases to get to the 15th. It would be much easier if there was a simpe name you could grep (even if it doesn't convey much information in many cases).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997)*

**Comment by knz** [comment]
Reviewed 2 of 2 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-Kp1HXKJ2SV70wJi2j9I:b-k7mrr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

An "either" direction is the same concept as the "exact match column" in `orderingInfo`. Perhaps we do want to return a more detailed answer in the form of an `orderingInfo`, but it's a bit hard to describe the semantics accurately. I will add a TODO for now.
</blockquote></details>

I still do not understand why the ordering doesn't matter here. Can you give an example?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-Kp1L4ajDerGnoGfd9CA:b-7v8log) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I still do not understand why the ordering doesn't matter here. Can you give an example?
</blockquote></details>

Either ordering will result in a successful merge-join. Because either the single value on the left is the same as the single value on the right, or we won't get any matches. In either case, it doesn't matter which way the merge-join orders rows w.r.t that column.

Examples (both examples have two equality columns and the first one is an exact match).

```
left          right
1 1           1 2
1 2           1 3
1 3           1 4
```
Here it won't matter which way the merge-join compares column `1`, it's all the same value.

```
left     right
1 1      2 2
1 2      2 3
1 3      2 4
```
It doesn't matter if inside the merge-join we first look at rows `1 x` and then `2 x` or vice-versa. There would be no results.

One could argue that this may matter as an optimization for outer joins if we prefer a certain output ordering, however I would counter-argue that for outer joins we can't really guarantee an ordering anyway because `NULL`s are interspersed with non-`NULL`s.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-Kp1NMTDx6LAQNeGJ7pq:b-xf4pr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Either ordering will result in a successful merge-join. Because either the single value on the left is the same as the single value on the right, or we won't get any matches. In either case, it doesn't matter which way the merge-join orders rows w.r.t that column.

Examples (both examples have two equality columns and the first one is an exact match).

```
left          right
1 1           1 2
1 2           1 3
1 3           1 4
```
Here it won't matter which way the merge-join compares column `1`, it's all the same value.

```
left     right
1 1      2 2
1 2      2 3
1 3      2 4
```
It doesn't matter if inside the merge-join we first look at rows `1 x` and then `2 x` or vice-versa. There would be no results.

One could argue that this may matter as an optimization for outer joins if we prefer a certain output ordering, however I would counter-argue that for outer joins we can't really guarantee an ordering anyway because `NULL`s are interspersed with non-`NULL`s.
</blockquote></details>

Is the meaning of "exact match" really "all the rows have the exact same value in that column"? Or rather "all the values are equal"?

I want to understand what happens if we have (e.g.) a decimal column with values 1.0, 1.000, 1.00

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-Kp1O73DM-sHMYKR9i6X:b-bb8t6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Is the meaning of "exact match" really "all the rows have the exact same value in that column"? Or rather "all the values are equal"?

I want to understand what happens if we have (e.g.) a decimal column with values 1.0, 1.000, 1.00

</blockquote></details>

Interesting question. I will add some notes on this in a comment in my next change which is changing `orderingInfo`.

Exact matches come from things like `WHERE x=1`, so I guess it means they are equal.
But I don't think it matters because the relative ordering of`1.0, 1.000, 1.00` does not matter so they can bet treated as the same value as far as orderings are concerned.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-Kp1Qeo-ENQ4QtoSjEnu:bpbjwu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Interesting question. I will add some notes on this in a comment in my next change which is changing `orderingInfo`.

Exact matches come from things like `WHERE x=1`, so I guess it means they are equal.
But I don't think it matters because the relative ordering of`1.0, 1.000, 1.00` does not matter so they can bet treated as the same value as far as orderings are concerned.
</blockquote></details>

The underlying problem, I think, is that if a column contains two representationally different values x and y such that "visually" (e.g. on pgwire) x != y but the SQL expression `x = y` is true, then ORDER BY on that column will be non-deterministic. 

I think this is OK but I would like to surface this in both our internal and external docs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997)*

**Comment by arjunravinarayan** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-Kp1S9coIJ3BEf_NGJ70:bkgyplu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The underlying problem, I think, is that if a column contains two representationally different values x and y such that "visually" (e.g. on pgwire) x != y but the SQL expression `x = y` is true, then ORDER BY on that column will be non-deterministic. 

I think this is OK but I would like to surface this in both our internal and external docs.
</blockquote></details>

I agree, I am considering taking that on - the intricacies of ordering is interesting enough that we should flesh it out in an internal document.

---

*[pkg/sql/ordering_test.go, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KotqlbayD6sqdYtZH8D:-Kp1RuIKg6J6Wj8ENIps:b-soml0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering_test.go#L275)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Excellent! BTW, this was also Jenny's idea originally (I forgot to credit).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997)*

**Comment by knz** [comment]
@arjunravinarayan please file an issue about the docs if you have an idea of what needs to be put into there. I'd like to have the issue tracked/linked from here.

LGTM otherwise.

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/ordering.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-KovvLwSeX0BbJDAIBnb:-Kp1U_EPaGqIzGPFQNXT:bgau5m6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ade0582fdb8d919abc4ec08698de72664b42bec/pkg/sql/ordering.go#L245)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

I agree, I am considering taking that on - the intricacies of ordering is interesting enough that we should flesh it out in an internal document.
</blockquote></details>

The alternative is to define an ordering even among those that are equal. E.g. collated strings could break ties by comparing the un-collated strings, etc. That would need us to adjust the definition of an "exact match column". But there are some complications here (like ensuring this ordering in an index).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997)*

**Comment by arjunravinarayan** [comment]
@knz: Done, #17028. Thank you for pressing on that. Let's continue this discussion there, so that we unblock the merging of this PR, which is otherwise good to ship. :shipit: 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16997#-:-Kp1k4-qtjYpaKKPOZaA:baoyp4a)*

**Comment by knz** [comment]
:+1: gtg
